tests/withBlockTracker.ts

```diff
@@ -0,0 +1,215 @@
+          });
+        } else if ('error' in stub.response) {
+          callback(null as unknown as Error, {
+            jsonrpc: '2.0',
+            id: 1,
+            error: {
+              code: -999,
+              message: stub.response.error,
+            },
+          });
+        }
+      } else if ('error' in stub) {
+        callback(new Error(stub.error), null as unknown as JsonRpcResponse<U>);
+      }
+      return;
+    }
+
+    callback(
+      new Error(
+        `Could not find any stubs matching "${request.method}". Perhaps they've already been called?\n\n` +
+          'The original set of stubs were:\n\n' +
+          `${util.inspect(this.originalStubs, { depth: null })}\n\n` +
+          'Current set of stubs:\n\n' +
+          `${util.inspect(this.stubs, { depth: null })}\n\n`,
+      ),
+      null as unknown as JsonRpcResponse<U>,
+    );
+  }
+
+  private buildStubsFrom(givenStubs: FakeProviderStub[]): FakeProviderStub[] {
+    const stubs = givenStubs.slice();
+
+    if (!stubs.some((stub) => stub.methodName === 'eth_blockNumber')) {
+      stubs.push({
+        methodName: 'eth_blockNumber',
+        response: {
+          result: '0x0',
+        },
+      });
+    }
+
+    if (!stubs.some((stub) => stub.methodName === 'eth_subscribe')) {
+      stubs.push({
+        methodName: 'eth_subscribe',
+        response: {
+          result: '0x0',
+        },
+      });
+    }
+
+    if (!stubs.some((stub) => stub.methodName === 'eth_unsubscribe')) {
+      stubs.push({
+        methodName: 'eth_unsubscribe',
+        response: {
+          result: true,
+        },
+      });
+    }
+
+    return stubs;
+  }
+}
+
+/**
+ * Calls the given function with the given subclass of BaseBlockTracker,
+ * ensuring that all listeners that are on the block tracker are removed and any
+ * timers or loops that are running within the block tracker are properly
+ * stopped.
+ *
+ * @param BlockTracker - A subclass of BaseBlockTracker.
+ * @param options - Options that allow configuring the block tracker or
+ * provider.
+ * @param callback - A callback which will be called with the built block
+ * tracker.
+ */
+async function withBlockTracker<T extends BlockTracker>(
+  BlockTracker: BlockTrackerConstructor<T>,
+  options: WithBlockTrackerOptions<T>,
+  callback: WithBlockTrackerCallback<T>,
+): Promise<void>;
+/**
+ * Calls the given function with the given subclass of BaseBlockTracker,
+ * ensuring that all listeners that are on the block tracker are removed and any
+ * timers or loops that are running within the block tracker are properly
+ * stopped.
+ *
+ * @param BlockTracker - A subclass of BaseBlockTracker.
+ * @param callback - A callback which will be called with the built instance of
+ * `BlockTracker`.
+ */
+async function withBlockTracker<T extends BlockTracker>(
+  BlockTracker: BlockTrackerConstructor<T>,
+  callback: WithBlockTrackerCallback<T>,
+): Promise<void>;
+/* eslint-disable-next-line jsdoc/require-jsdoc */
```

Comment:
We've had trouble getting types for overloaded functions working properly at all. I've been avoiding them. They seem easily avoided here as well, and it would make the function far simpler to use and to read.

Reply:
Unfortunately, if I collapse these, it doesn't seem that TypeScript is smart enough to figure out how to assign a type to a callback is passed as the final argument, and therefore it isn't able to assign types to the arguments that the callback takes. Perhaps this is fixed in a future version, but defining overloads seems to give it the help it needs.

Reply:
Hmm, that would clean this up a bit, but it would still be a variadic function, as the options bag would itself be optional (and optional arguments must be last). I guess I don't _need_ to do this, though — perhaps tests that don't need to pass in options can pass an empty object? Thought it would be sort of like passing `null` or `undefined`.

Reply:
I had forgotten that it was possible no options were needed. I see why this is variadic now. Accepting `null` or `undefined` as the first arg doesn't seem that bad to me, but I see why it's nice to avoid. I guess we'll see whether TypeScript can handle these overloads!

Reply:
I'm also curious to see how this turns out in VSCode and TypeDoc, but that can wait for another time.